Realm/Realm/Handles/SessionHandle.cs

```diff
@@ -224,14 +261,86 @@ private static void HandleSessionError(IntPtr sessionHandlePtr, ErrorCode errorC
                     exception = new SessionException(messageString, errorCode);
                 }
 
-                Session.RaiseError(session, exception);
+                if (syncConfigurationBase.OnSessionError != null)
+                {
+                    syncConfigurationBase.OnSessionError.Invoke(session, exception);
+                }
+                else
+                {
+                    // after deprecation: this will need to go when Session.Error is fully deprecated
+                    Session.RaiseError(session, exception);
+                }
             }
             catch (Exception ex)
             {
                 Logger.Default.Log(LogLevel.Warn, $"An error has occurred while handling a session error: {ex}");
             }
         }
 
+        [MonoPInvokeCallback(typeof(NativeMethods.NotifyBeforeClientReset))]
+        private static bool NotifyBeforeClientReset(IntPtr beforeFrozen, IntPtr managedSyncConfigurationHandle)
+        {
+            var syncConfigHandle = GCHandle.FromIntPtr(managedSyncConfigurationHandle);
+            var syncConfiguration = (SyncConfigurationBase)syncConfigHandle.Target;
+
+            if (syncConfiguration.ClientResetHandler == null)
+            {
+                return true;
+            }
+
+            var discardLocalResetHandler = (DiscardLocalResetHandler)syncConfiguration.ClientResetHandler;
+            var schema = syncConfiguration.GetSchema();
+
+            try
```

Comment:
1. A wrong cast should not happen as the types are checked way ahead of time, [here](https://github.com/realm/realm-dotnet/blob/582cd17185e7031756863d60f41508c52d93138c/Realm/Realm/Configurations/SyncConfigurationBase.cs#L139). So if we got here the type must be right.
1. I think that a wrong cast would be an error on our side, which the tests will spot right away.
1. If your concern is the fact that the user may change on the fly the type of handler, since they are mutable; that's a fair point. And I would not know what the best approach is. Likely we should not make the application go down, although shall we not tell the user that what he is doing is far from correct? This is different from a logical error that they could make in their handler. In fact the logical error may show up only in corner cases so I don't think we should make the app crash. And in fact, we rightly don't. But a change of handler type on the fly is a consistent error that should show up any time that code path is reached. Because of this, I assume that these types of mistakes are easily findable by testing that reaches that code path. Shall we assume that users don't do testing? But if that is the case, I don't know who's at fault there.
Unfortunately, this brings up another problem. How can we tell the user that changing callback, without modifying its type but only its content, after the instance of realm is created won't do what is expected?

In summary if you referred to point 3, I think we should discuss it a little further. Otherwise, I must have missed something that didn't make me understand your concern.

Reply:
My concern is precisely 3. - if the user does:

```csharp
var config = new PartitionSyncConfiguration
{
    ClientResetHandler = new DiscardLocalHandler()
};

var realm = Realm.GetInstance(config);
config.ClientResetHandler = new ManualRecoveryHandler();
```

The application would crash with the code above and not necessarily in an obvious way as in native, we won't be able to catch the managed exception that we will throw here. Moving it inside the `try`-block ensures that we log the error and that the developer is notified of what they did wrong.